### PR TITLE
[FEA] Support BlazingSQL distributed queries w/ UCX comms, add example using node cluster module

### PR DESCRIPTION
**Pull request by mzegar**

### CODE REVIEW

File: modules/blazingsql/cluster.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+const createTable = 'createTable';
+const runQuery = 'runQuery';
+const queryRan = 'queryRan';
+
+const numberOfWorkers = 2;
+
+const ucp_context = new UcpContext();
+
+if (cluster.isMaster) {
```
Comment (trxcllnt) [inline modules/blazingsql/cluster.js:13]: https://nodejs.org/api/child_process.html#child_process_advanced_serialization

```suggestion
if (cluster.isMaster) {
  cluster.setupMaster({ serialization: 'advanced' });
```

File: modules/blazingsql/cluster.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      ucpContext: ucp_context,
+    };
+  }));
+
+  const a = Series.new([1, 2, 3]);
+  const b = Series.new([1, 2, 3]);
+  const df = new DataFrame({ 'a': a, 'b': b });
+
+  workers.forEach((w) => {
+    w.send({ operation: createTable, tableName: 'test_table', dataframe: df });
```
Comment (trxcllnt) [inline modules/blazingsql/cluster.js:33]: I think we'll need to send the table in binary form here, then we can deserialize it on the child side:

```suggestion
    w.send({
    operation: createTable,
    tableName: 'test_table',
    dataframe: df.toArrow().serialize()
    });
```

File: modules/blazingsql/cluster.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      ip: 'localhost',
+      port: 8000,
+      ucpContext: ucp_context,
+    }
+  ]);
+
+  process.on('message', (args) => {
+    if (args.operation === createTable) {
+      console.log(`Creating table: ${args.tableName}`);
+      bc.createTable(args.tableName, args.dataframe);
```
Comment (trxcllnt) [inline modules/blazingsql/cluster.js:67]:
```suggestion
      bc.createTable(args.tableName, DataFrame.fromArrow(args.dataframe));
```

File: modules/blazingsql/src/blazingcontext.ts [hunk 4]
```diff
@@ -42,7 +43,7 @@ export class BlazingContext {
   private generator: any;
   private tables: Map<string, DataFrame>;
 
-  constructor() {
+  constructor(workers: WorkerUcpInfo[] = []) {
```
Comment (trxcllnt) [inline modules/blazingsql/src/blazingcontext.ts:46]: I believe here we'll need the ability to control all the configuration options passed to `new Context()` below.

File: modules/blazingsql/cluster.js (hunk 2 above)
Comment (mzegar) [reply modules/blazingsql/cluster.js:33]: Thanks, yeah, was going to poke you abot this on monday.

File: modules/blazingsql/cluster.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+} else if (cluster.isWorker) {
+
+  const bc = new BlazingContext([
+    {
+      workerId: cluster.worker,
+      ip: 'localhost',
+      port: 8000,
+      ucpContext: ucp_context,
+    }
+  ]);
```
Comment (trxcllnt) [inline modules/blazingsql/cluster.js:71]: Doesn't this list need to include an object for all workers?

File: modules/blazingsql/wrapper/blazingsql/cache.cpp [hunk 6]
```diff
@@ -31,6 +31,21 @@ CacheMachine::wrapper_t CacheMachine::New(Napi::Env const& env,
   return inst;
 }
 
+void CacheMachine::add_to_cache(std::string const& message_id,
+                                std::vector<std::string> const& table_names,
+                                cudf::table_view const& table_view) {
+  std::unique_ptr<ral::frame::BlazingTable> table =
+    std::make_unique<ral::frame::BlazingTable>(table_view, table_names);
+
+  this->_cache->addToCache(std::move(table), "", true, {}, true);
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/cache.cpp:40]:
```suggestion
  this->_cache->addToCache(std::move(table), message_id, true, {}, true);
```

File: modules/blazingsql/wrapper/blazingsql/cache.hpp [hunk 7]
```diff
@@ -48,6 +49,13 @@ struct CacheMachine : public nv::EnvLocalObjectWrap<CacheMachine> {
 
   inline operator std::shared_ptr<ral::cache::CacheMachine>() { return _cache; }
 
+  void add_to_cache(std::string const& message_id,
+                    std::vector<std::string> const& table_names,
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/cache.hpp:53]:
```suggestion
                    std::vector<std::string> const& column_names,
```

File: modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp [hunk 8]
```diff
@@ -24,14 +25,29 @@ ContextWrapper::wrapper_t ContextWrapper::New(
...
 }
 
 ContextWrapper::ContextWrapper(Napi::CallbackInfo const& info)
   : EnvLocalObjectWrap<ContextWrapper>(info) {}
 
+void ContextWrapper::add_to_cache(std::string const& message_id,
+                                  std::vector<std::string> const& table_names,
+                                  cudf::table_view const& table_view) {
+  this->_transport_in.Value()->add_to_cache(message_id, table_names, table_view);
+}
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp:46]:
```suggestion
void ContextWrapper::add_to_cache(std::string const& message_id,
                                  std::vector<std::string> const& column_names,
                                  cudf::table_view const& table_view) {
  this->_transport_in.Value()->add_to_cache(message_id, column_names, table_view);
}
```

File: modules/blazingsql/wrapper/blazingsql/contextwrapper.hpp [hunk 9]
```diff
@@ -38,17 +39,26 @@ struct ContextWrapper : public EnvLocalObjectWrap<ContextWrapper> {
...
+                                 int> pair,
+                       Napi::Object const& ucp_context);
 
   /**
    * @brief Construct a new ContextWrapper instance from JavaScript.
    */
   ContextWrapper(Napi::CallbackInfo const& info);
 
+  void add_to_cache(std::string const& message_id,
+                    std::vector<std::string> const& table_names,
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/contextwrapper.hpp:51]:
```suggestion
                    std::vector<std::string> const& column_names,
```

File: modules/blazingsql/wrapper/blazingsql/cache.cpp [hunk 10]
```diff
@@ -31,6 +31,21 @@ CacheMachine::wrapper_t CacheMachine::New(Napi::Env const& env,
   return inst;
 }
 
+void CacheMachine::add_to_cache(std::string const& message_id,
+                                std::vector<std::string> const& table_names,
+                                cudf::table_view const& table_view) {
+  std::unique_ptr<ral::frame::BlazingTable> table =
+    std::make_unique<ral::frame::BlazingTable>(table_view, table_names);
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/cache.cpp:38]:
```suggestion
void CacheMachine::add_to_cache(std::string const& message_id,
                                std::vector<std::string> const& column_names,
                                cudf::table_view const& table_view) {
  std::unique_ptr<ral::frame::BlazingTable> table =
    std::make_unique<ral::frame::BlazingTable>(table_view, column_names);
```

File: modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp [hunk 11]
```diff
@@ -24,14 +25,29 @@ ContextWrapper::wrapper_t ContextWrapper::New(
...
   return inst;
 }
 
 ContextWrapper::ContextWrapper(Napi::CallbackInfo const& info)
   : EnvLocalObjectWrap<ContextWrapper>(info) {}
 
+void ContextWrapper::add_to_cache(std::string const& message_id,
+                                  std::vector<std::string> const& column_names,
+                                  cudf::table_view const& table_view) {
+  this->_transport_in.Value()->add_to_cache(message_id, column_names, table_view);
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp:45]:
```suggestion
  this->_transport_out.Value()->add_to_cache(message_id, column_names, table_view);
```

File: modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp [hunk 12]
```diff
@@ -24,14 +25,29 @@ ContextWrapper::wrapper_t ContextWrapper::New(
...
+}
+
+std::tuple<std::vector<std::string>, std::unique_ptr<cudf::table>> ContextWrapper::pull_from_cache(
+  std::string const& message_id) {
+  return this->_transport_out.Value()->pull_from_cache(message_id);
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp:50]:
```suggestion
  return this->_transport_in.Value()->pull_from_cache(message_id);
```

File: modules/blazingsql/wrapper/blazingsql/api.cpp [hunk 13]
```diff
@@ -13,39 +13,67 @@
...
+        } else if (config[name] == "false") {
+          config[name] = "False";
+        }
       }
     }
     return config;
   }();
 
+  Napi::Array objects               = props.Get("workersUcpInfo");
+  UcpContext::wrapper_t ucp_context = UcpContext::wrapper_t();
```
Comment (mzegar) [inline modules/blazingsql/wrapper/blazingsql/api.cpp:55]: I think the only thing I'd need a suggestion on how to handle, we create a `ucp_context` for all cases, should we avoid creating it if we are doing only single node?

Comment (trxcllnt) [reply modules/blazingsql/wrapper/blazingsql/api.cpp:55]: This is constructing an empty `Napi::Object` instance, which should be fine as long as we don't try to use it when `singleNode=true`.

File: modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp [hunk 14]
```diff
@@ -13,25 +13,55 @@
...
+                                  std::string const& message_id,
+                                  std::vector<std::string> const& column_names,
+                                  cudf::table_view const& table_view) {
+  if (src_ral_id == dst_ral_id) {
+    this->_transport_in.Value()->add_to_cache(
+      node_id, src_ral_id, dst_ral_id, ctx_token, message_id, column_names, table_view);
+  } else {
+    this->_transport_out.Value()->add_to_cache(
+      node_id, src_ral_id, dst_ral_id, ctx_token, message_id, column_names, table_view);
+  }
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp:59]: I forgot to delete this after testing -- a BlazingContext can't send messages to itself.

```suggestion
  this->_transport_out.Value()->add_to_cache(
    node_id, src_ral_id, dst_ral_id, ctx_token, message_id, column_names, table_view);
```

File: modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp [hunk 15]
```diff
@@ -13,25 +13,55 @@
...
   Napi::Env const& env,
+  int32_t const& ral_id,
   std::pair<
     std::pair<std::shared_ptr<ral::cache::CacheMachine>, std::shared_ptr<ral::cache::CacheMachine>>,
-    int> args) {
-  auto inst    = EnvLocalObjectWrap<ContextWrapper>::New(env, {});
-  auto& caches = args.first;
-  inst->_port  = args.second;
+    int> const& pair,
+  Napi::Object const& ucp_context) {
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/contextwrapper.cpp:32]: We can strongly-type this argument:

```suggestion
  UcpContext::wrapper_t const& ucp_context) {
```

File: modules/blazingsql/wrapper/blazingsql/contextwrapper.hpp [hunk 16]
```diff
@@ -28,27 +29,44 @@ struct ContextWrapper : public EnvLocalObjectWrap<ContextWrapper> {
...
    *
    * @return wrapper_t The new ContextWrapper instance
    */
   static wrapper_t New(Napi::Env const& env,
+                       int32_t const& ral_id,
                        std::pair<std::pair<std::shared_ptr<ral::cache::CacheMachine>,
                                            std::shared_ptr<ral::cache::CacheMachine>>,
-                                 int> args);
+                                 int> const& pair,
+                       Napi::Object const& ucp_context);
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/contextwrapper.hpp:44]:
```suggestion
                       UcpContext::wrapper_t const& ucp_context);
```

File: modules/blazingsql/wrapper/blazingsql/contextwrapper.hpp [hunk 17]
```diff
@@ -28,27 +29,44 @@ struct ContextWrapper : public EnvLocalObjectWrap<ContextWrapper> {
...
+
+  std::tuple<std::vector<std::string>, std::unique_ptr<cudf::table>> pull_from_cache(
+    std::string const& message_id);
+
  private:
   int32_t _port{};
+  int32_t _ral_id{};
   Napi::Reference<Wrapper<CacheMachine>> _transport_out;
   Napi::Reference<Wrapper<CacheMachine>> _transport_in;
+  Napi::ObjectReference _ucp_context;
```
Comment (trxcllnt) [inline modules/blazingsql/wrapper/blazingsql/contextwrapper.hpp:69]:
```suggestion
  Napi::Reference<UcpContext::wrapper_t> _ucp_context;
```

File: modules/blazingsql/wrapper/blazingsql/api.cpp (hunk 13 above)
Comment (mzegar) [reply modules/blazingsql/wrapper/blazingsql/api.cpp:55]: ah right, should be fine then.

**Review by trxcllnt — APPROVED** [review]

**Pull request merged**